### ISSUE TITLE
Texture is distorted in Line2D when width_curve applied

### ISSUE THREAD

**Issue by timcode77** [godotengine/godot#83306]
### Godot version

v4.1.2.stable.official [399c9dc39]

### System information

Windows 11 Home  Vulkan API 1.2.170 - Forward+ - Using Vulkan Device #0: Intel - Intel(R) UHD Graphics

### Issue description

Textures are distorted in a strange way when applied as a fill on a Line2D with a width_curve applied.

![image](https://github.com/godotengine/godot/assets/82291606/8312c771-9c5b-4a53-981f-ff6ed9ceaaf5)

### Steps to reproduce

1. Create a Line2D of any length with 2 points.
2. Create a linear width_curve for the Line2D.
3. Apply a texture fill.

### Minimal reproduction project

[line2D_texture_bug.zip](https://github.com/godotengine/godot/files/12902737/line2D_texture_bug.zip)

**Comment by MewPurPur** [comment]
With the way Line2D is built, this is expected.

![image](https://github.com/godotengine/godot/assets/85438892/b6ca20fc-c3cb-4665-ba5a-7954a6590ae0)

Maybe we can come up with some better way to do UVs but idk about that.

**Comment by MewPurPur** [comment]
https://www.cs.ucdavis.edu/~amenta/s12/perspectiveCorrect.pdf

https://mtrebi.github.io/2017/03/15/texture-mapping-affine-perspective.html

Well this might help if someone tries to fix the UVs of this node.

**Comment by timcode77** [comment]
I would love to help but alas I have never contributed to the code base and my C++ skills were last put into action about 20 years ago.

One possibility is that if a new algorithm is implemented, it is created as an option to provide backward compatibility. 

It could perhaps appear beneath "Texture Mode" in the inspector.

Something like:

**Texture Mapping Mode**:
- **Affine** (backward compatibility mode)
- **Interpolated** ("perspective-correct" mode) 
 
I borrowed some language from @MewPurPur's first link above (https://www.cs.ucdavis.edu/~amenta/s12/perspectiveCorrect.pdf).

**Comment by MewPurPur** [comment]
IMO we can count this as a bug and we don't _need to_ make the solution more complex by providing compatibility.

**Comment by timcode77** [comment]
> IMO we can count this as a bug and we don't _need to_ make the solution more complex by providing compatibility. Like we don't need to provide a "is_broken" property every time we fix something some users might use intentionally 😄

Hopefully we get a nice interpolated ("perspective-correct") texture algorithm sometime!

**Comment by Calinou** [comment]
@timcode77 Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by TheSofox** [comment]
Been looking into this. `Line2D` uses `LineBuilder` to generate the actual indices, vertices, colors and UVs to use for generating the line. This data is then given to `RendererCanvasCull::canvas_item_add_triangle_array` (a class used for rendering CanvasItems) and then on to create the raw polygons that will be rendered.

The things is, all this data is correct. The UVs are for each individual vertex and are perfectly fine, the issue is interpolating these UVs across the polygon. The problem is I have no idea where such interpolation is done. If we were to create a new algorithm for this, I have no idea where in the source code it should be placed.

**Comment by Calinou** [comment]
> the issue is interpolating these UVs across the polygon. The problem is I have no idea where such interpolation is done.

The interpolation can either be done by using a custom shader for the polygons used in Line2D drawing, or creating additional vertices that have adjusted UVs in between points (kind of like how subdividing meshes in PS1 games reduced affine texture warping). I'd prefer the latter solution personally, as it doesn't involve a custom shader that would break user-provided shaders.

**Comment by MewPurPur** [comment]
I don't think so, you can have this issue even if the Line2D segment has a trapezoid-like area between the joints.

Line2D is a messy node that's been bloated with features and implementation details that don't play well together ever since it's implementation 8 years ago. I feel like shaders are the better solution, given that we also need them to truly solve Line2D gradients (#50076). Otherwise, for both issues, we can't do any better than a half solution.

**Comment by TheSofox** [comment]
Okay, are we looking for a quick fix here or a full `Line2d` rewrite?

**Comment by xolarkodak** [comment]
Hello, I also got tired of this artifact and decided to figure out its cause. The problem is **not** in the implementation of `line_builder.cpp`, nor in incorrect geometry construction - it lies in the incorrect UV calculations (interpolation) for the mesh. This artifact will also appear if you start building trajectories using Polygon2D or in any mesh.

![Image](https://github.com/user-attachments/assets/c4b8fe8e-185d-466b-91c6-04ed7f8979d6)
_Here is a Polygon2D with the same artifact._

That makes sense, since all line_builder.cpp does is generate a mesh - and Polygon2D is also a mesh.

This is a well-known classic computer graphic problem that appears everywhere: from _Unity_ and _Blender_ to _OpenGL_. It all stems from **linear UV interpolation** after a mesh is split into triangles.

Don’t look at it as a trapezoid - look at it as two triangles, and you’ll understand the behavior and nature of the texture distortion.

https://github.com/user-attachments/assets/31491504-2fe4-4c4e-9a68-548f694c4aca

![Image](https://github.com/user-attachments/assets/686206dd-909a-449a-9633-f34870561fc3)

It is necessary to develop a projection correction shader for the canvas in 2D, if possible. Help and advice from a more experienced computer graphics developer is needed.

Here are some links to similar issues, where you can read more explanations, etc:
- https://godotforums.org/d/34677-uv-on-generated-meshes
- https://stackoverflow.com/questions/15242507/perspective-correct-texturing-of-trapezoid-in-opengl-es-2-0
- https://blender.stackexchange.com/questions/156395/trapezopid-shaped-face-to-square-uv-coordinates-has-unfavourable-skewing
- https://discussions.unity.com/t/how-to-smooth-uv-in-mesh/632106
- https://gamedev.stackexchange.com/questions/148082/how-can-i-fix-zig-zagging-uv-mapping-artifacts-on-a-generated-mesh-that-tapers
- https://gamedev.stackexchange.com/questions/174857/mapping-a-texture-to-a-2d-quadrilateral/174871
